File: app/src/main/java/de/grobox/transportr/utils/DateUtils.java
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2018 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.utils;

import android.content.Context;
import android.support.annotation.Nullable;
import android.widget.TextView;

import java.text.DateFormat;
import java.util.Calendar;
import java.util.Date;

import de.grobox.transportr.R;

import static android.text.format.DateUtils.MINUTE_IN_MILLIS;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;

public class DateUtils {

	public static String getDate(Context context, Date date) {
		DateFormat df = android.text.format.DateFormat.getDateFormat(context);

		return df.format(date);
	}

	public static String getTime(Context context, @Nullable Date date) {
		if (date == null) return "";
		DateFormat tf = android.text.format.DateFormat.getTimeFormat(context);

		if (tf.getNumberFormat().getMinimumIntegerDigits() == 1) {
			String formattedTime = tf.format(date);
			if (formattedTime.indexOf(':') == 1) {
				// ensure times always have the same length, so views (like intermediate stops) align
				return "0" + formattedTime;
			}
			return formattedTime;
		}
		return tf.format(date);
	}

	public static String getDuration(long duration) {
		// get duration in minutes
		duration = duration / 1000 / 60;

		long m = duration % 60;
		long h = duration / 60;

		return Long.toString(h) + ":" + (m < 10 ? "0" : "") + Long.toString(m);
	}

	public static String getDuration(Date start, Date end) {
		return getDuration(end.getTime() - start.getTime());
	}

	/**
	 * Returns delay in minutes
	 */
	public static long getDelay(Date plannedTime, Date predictedTime) {
		return (predictedTime.getTime() - plannedTime.getTime()) / 1000 / 60;
	}

	public static String getDelayString(long delay) {
		return (delay > 0 ? "+" : "") + Long.toString(delay);
	}

	@Nullable
	public static String getDelayText(long delay) {
		if (delay >= 0) {
			return "+" + Long.toString(delay / 1000 / 60);
		} else if (delay < 0) {
			return "-" + Long.toString(delay / 1000 / 60);
		} else {
			return null;
		}
	}

	public static boolean isToday(Calendar calendar) {
		return android.text.format.DateUtils.isToday(calendar.getTimeInMillis());
	}

	public static boolean isNow(Calendar calendar) {
		long diff = Math.abs(calendar.getTimeInMillis() - Calendar.getInstance().getTimeInMillis());
		return diff < 10 * MINUTE_IN_MILLIS;
	}

	public static void setRelativeDepartureTime(TextView timeRel, Date date) {
		long difference = getDifferenceInMinutes(date);
		if (difference > 99 || difference < -99) {
			timeRel.setVisibility(GONE);
		} else if (difference == 0) {
			timeRel.setText(timeRel.getContext().getString(R.string.now_small));
			timeRel.setVisibility(VISIBLE);
		} else if (difference > 0) {
			timeRel.setText(timeRel.getContext().getString(R.string.in_x_minutes, difference));
			timeRel.setVisibility(VISIBLE);
		} else {
			timeRel.setText(timeRel.getContext().getString(R.string.x_minutes_ago, difference * -1));
			timeRel.setVisibility(VISIBLE);
		}
	}

	private static long getDifferenceInMinutes(Date date) {
		return getDelay(new Date(), date);
	}

}

File: app/src/main/java/de/grobox/transportr/utils/DateUtils.kt
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2021 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package de.grobox.transportr.utils

import android.content.Context
import android.text.format.DateFormat
import android.text.format.DateUtils
import android.view.View
import androidx.core.content.ContextCompat
import de.grobox.transportr.R
import java.util.*
import kotlin.math.abs

object DateUtils {
    fun millisToMinutes(millis: Long): Long {
        val seconds = millis / 1000
        return seconds / 60 + when {
            seconds % 60 >= 30 -> 1
            seconds % 60 <= -30 -> -1
            else -> 0
        }
    }

    fun formatDate(context: Context, date: Date): String {
        val df = DateFormat.getDateFormat(context)
        return df.format(date)
    }

    fun formatTime(context: Context, date: Date?): String {
        if (date == null) return ""
        val tf = DateFormat.getTimeFormat(context)
        if (tf.numberFormat.minimumIntegerDigits == 1) {
            val formattedTime = tf.format(date)
            return if (formattedTime.indexOf(':') == 1) {
                // ensure times always have the same length, so views (like intermediate stops) align
                "0$formattedTime"
            } else formattedTime
        }
        return tf.format(date)
    }

    fun formatDuration(duration: Long): String {
        // get duration in minutes
        val durationMinutes = millisToMinutes(duration)
        val m = durationMinutes % 60
        val h = durationMinutes / 60
        return "$h:${m.toString().padStart(2, '0')}"
    }

    fun formatDuration(start: Date, end: Date): String {
        return formatDuration(end.time - start.time)
    }

    fun formatDelay(context: Context, delay: Long): Delay {
        val delayMinutes = millisToMinutes(delay)
        return Delay(
            delay = "${if (delayMinutes >= 0) '+' else ""}$delayMinutes",
            color = ContextCompat.getColor(context, if (delayMinutes > 0) R.color.md_red_500 else R.color.md_green_500)
        )
    }

    data class Delay(
        val delay: String,
        val color: Int
    )

    fun isToday(calendar: Calendar): Boolean {
        return DateUtils.isToday(calendar.timeInMillis)
    }

    fun isWithinMinutes(calendar: Calendar, minutes: Int): Boolean {
        val diff = abs(calendar.timeInMillis - Calendar.getInstance().timeInMillis)
        return diff < minutes * DateUtils.MINUTE_IN_MILLIS
    }

    fun isNow(calendar: Calendar): Boolean = isWithinMinutes(calendar, 1)

    fun formatRelativeTime(context: Context, date: Date, max: Int = 99): RelativeTime {
        val difference = getDifferenceInMinutes(date)
        return RelativeTime(
            relativeTime = when {
                difference !in -max..max -> ""
                difference == 0L -> context.getString(R.string.now_small)
                difference > 0 -> context.getString(R.string.in_x_minutes, difference)
                else -> context.getString(R.string.x_minutes_ago, difference * -1)
            },
            visibility = if (difference in -max..max) View.VISIBLE else View.GONE
        )
    }

    data class RelativeTime(
        val relativeTime: String,
        val visibility: Int
    )

    /**
     * Returns difference in minutes
     */
    private fun getDifferenceInMinutes(d1: Date, d2: Date): Long {
        return millisToMinutes(d2.time - d1.time)
    }

    private fun getDifferenceInMinutes(date: Date): Long {
        return getDifferenceInMinutes(Date(), date)
    }
}